rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+# Distributed tf.data service
+
+| Status        | Proposed          |
+| :------------ | :------------------------------------------------------ |
+| **RFC #**     | [195](https://github.com/tensorflow/community/pull/195) |
+| **Author(s)** | Andrew Audibert (aaudibert@google.com) Rohan Jain (rohanj@google.com) |
+| **Sponsor**   | Jiri Simsa (jsimsa@google.com)                          |
+| **Updated**   | 2019-01-09                                              |
+
+## Objective
+
+Provide an API and implementation of a tf.data service which can process tf.data
+datasets in a distributed manner. The service can be run outside the TensorFlow
+cluster or be exported as a gRPC service by TensorFlow servers.
+
+Goals:
+
+-   Enable horizontal scaling of dataset computation to improve performance of
+    input-bound dataset pipelines.
+-   Improve tf.data integration with the tf.distribute API. In particular,
+    support dynamic sharding of data across multiple processes.
+-   Provide visitation guarantees for distributed training jobs.
+
+Non-goals:
+
+-   Process non-dataset data.
+-   Distribute datasets that rely on external / non-serializable state.
+-   Support non-graph computation (e.g. py_function).
+
+## Motivation
+
+### Host machine input pipelines can't always keep up with accelerators.
+
+Some input pipelines require significant resources to produce their data, e.g.
+due to image transformations. When the host machine isn't powerful enough to
+generate input data at the rate the attached accelerator(s) consume the data,
+the accelerator(s) will idle. This slows down training time, and also wastes
+valuable accelerator resources. The tf.data service solves this problem by using
+N input workers to feed M accelerators. The number of input workers can be
+scaled up or down as needed to keep up with the accelerators.
+
+### Distributed training requires a distribution-aware input pipeline.
+
+Today tf.data supports the tf.distribute API by providing mechanisms for
+sharding, cloning, and re-batching. The tf.distribute API uses these primitives
+to implement their own version of a distributed dataset. If distributed datasets
+become a core feature of tf.data, tf.data can provide a public API for
+tf.distribute (and users who wish to implement their own distribution) to use
```

Comment:
This is really important, and I'd like to review the distributed data APIs independently from the service model.

Reply:
Does the "User-facing Python API" section address this?